File: Challenges/MaxAreaOfIsland.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        Queue<ArrayList<Integer>> q = new LinkedList<>();
        for(int a=0; a<grid.length; a++){
            for(int b=0; b<grid[0].length; b++){
                if (grid[a][b] == 1){
                    q.add(new ArrayList<Integer>(Arrays.asList(a, b)));
                }
                int area = 0;
                while(q.peek() != null){
                    List<Integer> curr = q.poll();
                    int i = curr.get(0);
                    int j = curr.get(1);
                    grid[i][j] = -1;
                    area += 1;
                    if (i-1>=0 && grid[i-1][j] == 1 && !q.contains(new ArrayList<Integer>(Arrays.asList(i-1, j)))){
                        q.add(new ArrayList<Integer>(Arrays.asList(i-1, j)));
                    }
                    if (i+1<grid.length && grid[i+1][j] == 1 && !q.contains(new ArrayList<Integer>(Arrays.asList(i+1, j)))){
                        q.add(new ArrayList<Integer>(Arrays.asList(i+1, j)));
                    }
                    if (j-1>=0 && grid[i][j-1] == 1 && !q.contains(new ArrayList<Integer>(Arrays.asList(i, j-1)))){
                        q.add(new ArrayList<Integer>(Arrays.asList(i, j-1)));
                    }
                    if (j+1<grid[0].length && grid[i][j+1] == 1 && !q.contains(new ArrayList<Integer>(Arrays.asList(i, j+1)))){
                        q.add(new ArrayList<Integer>(Arrays.asList(i, j+1)));
                    }
                }
                maxArea = Math.max(maxArea, area);
            }
        }
        return maxArea;
    }
}

class SolutionDFS {
    public int maxAreaOfIsland(int[][] grid) {
        int max = 0;
        
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    max = Math.max(dfs(grid, i, j), max);
                }
            }
        }
        
        return max;
    }
    
    int dfs(int[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == 0) {
            return 0;
        }
        
        if (grid[i][j] == 1) {
            grid[i][j] = 999;
            return 1 + dfs(grid, i + 1, j) + dfs(grid, i - 1, j) + dfs(grid, i, j + 1) + dfs(grid, i, j - 1) ;
         }
        return 0;
    }
}

File: Challenges/maxAreaOfIsland.py
"""
Max Area of Island

Given a non-empty 2D array grid of 0's and 1's, an island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.

Find the maximum area of an island in the given 2D array. (If there is no island, the maximum area is 0.)

Example 1:

[[0,0,1,0,0,0,0,1,0,0,0,0,0],
 [0,0,0,0,0,0,0,1,1,1,0,0,0],
 [0,1,1,0,1,0,0,0,0,0,0,0,0],
 [0,1,0,0,1,1,0,0,1,0,1,0,0],
 [0,1,0,0,1,1,0,0,1,1,1,0,0],
 [0,0,0,0,0,0,0,0,0,0,1,0,0],
 [0,0,0,0,0,0,0,1,1,1,0,0,0],
 [0,0,0,0,0,0,0,1,1,0,0,0,0]]
Given the above grid, return 6. Note the answer is not 11, because the island must be connected 4-directionally.
Example 2:

[[0,0,0,0,0,0,0,0]]
Given the above grid, return 0.
Note: The length of each dimension in the given grid does not exceed 50.
"""



"""
DFS
Time: O(MN)
Space: O(MN)
"""

class Solution:
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        maxArea, self.area= 0,0
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                if grid[x][y] == 1:
                    self.getArea(grid,(x,y))
                    maxArea = max(maxArea,self.area)
                    self.area=0
        return maxArea
                    
    def getArea(self,grid,point):
        self.area+=1
        x,y = point
        grid[x][y]=0
        for u,v in [(1,0),(-1,0),(0,1),(0,-1)]:
            if x+u >= 0 and x+u  < len(grid):
                if y+v >= 0 and y+v < len(grid[0]):
                    if grid[x+u][y+v]==1:
                        self.getArea(grid,(x+u,y+v))


# iterative
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        # DFS
        if len(grid) == 0:
            return 0
        height = len(grid)
        width = len(grid[0])
        visit = set()
        max_area = 0
        for i in range(height):
            for j in range(width):
                if grid[i][j] == 1 and (i,j) not in visit:
                    visit.add((i,j))
                    stack = [(i,j)]
                    area = 1
                    while stack:
                        u = stack.pop()
                        i,j = u[0],u[1]
                        if i > 0 and grid[i-1][j] == 1 and (i-1,j) not in visit:
                            stack.append((i-1,j))
                            visit.add((i-1,j))
                            area += 1
                        if i < height - 1 and grid[i+1][j] == 1 and (i+1,j) not in visit:
                            stack.append((i+1,j))
                            visit.add((i+1,j))
                            area += 1
                        if j > 0 and grid[i][j-1] == 1 and (i,j-1) not in visit:
                            stack.append((i,j-1))
                            visit.add((i,j-1))
                            area += 1
                        if j < width - 1 and grid[i][j+1] == 1 and (i,j+1) not in visit:
                            stack.append((i,j+1))
                            visit.add((i,j+1))
                            area += 1
                    if area > max_area:
                        max_area = area
        return max_area
